### ISSUE TITLE
kubelet leaves pod in existence when it no longer exists due to devicemapper bug

### ISSUE THREAD

**Issue by andrewmichaelsmith** [kubernetes/kubernetes#24252]
Verison: 1.2.0 / Docker 1.8.3

We have come across a (so far undiagnosed) issue with docker/devicemapper (https://github.com/docker/docker/issues/22031) where by the Docker API returns a HTTP 500 when the kubelet polls it for a particular pod:

```Apr 14 13:25:33 server kubelet[14718]: E0414 13:25:33.159976   14718 pod_workers.go:138] Error syncing pod 8f6b7818-00c7-11e6-8eb8-00163e1b3b80, skipping: API error (500): Unknown device 91ac35b4dad8ad64abc7a8f58c22f48b6b9533f8aa81bf3138152ac52eb23163```

If we delete this pod it results results in the pod being in a "Terminating" state:

```Events:
  FirstSeen     LastSeen        Count   From                            SubobjectPath   Type            Reason          Message
  ---------     --------        -----   ----                            -------------   --------        ------          -------
  1d            10s             12749   {kubelet server}                  Warning         FailedSync      Error syncing pod, skipping: API error (500): Unknown device 91ac35b4dad8ad64abc7a8f58c22f48b6b9533f8aa81bf3138152ac52eb23163
```

The problem is that this will stay forever (you can see that this is one day, we've seen upwards of a week). We've found two cases with two different fixes:

* 1 - Container shows in `docker ps -a` as `Created`: Fix is to delete the container from there
* 2 - Container doesn't show anywhere

For 1) if we delete the container the kubelet recognises it's gone and "removes" the pod. For 2) the only fix I've found is to restart the kubelet.

I appreciate this is a bit tricky - it looks like a docker/devicemaper bug. And I'm not quite sure what the fix would be - right now from our point of view in the evnet of a 500 the kubelet should delete the Pod. This would solve our problems. Of course I'm not clear of other implications this may have.

It does seem to be (especially for point 2)) that there should perhaps be some timeout on knowledge about containers. If you polled for a full list of containers `91ac35b4dad8ad64abc7a8f58c22f48b6b9533f8aa81bf3138152ac52eb23163` would not be there. The problem is that this never happens (or if it does the data isn't being used in relation to this pod).

**Comment by ncdc** [comment]
cc @derekwaynecarr @rhvgoyal

I get the "unknown device" errors quite frequently in my development environment, but they haven't resulted in pods stuck Terminating (or I haven't noticed, perhaps).

**Comment by yujuhong** [comment]
@andrewmichaelsmith, a temporary workaround (which I think should work) is to forcefully delete the pod `kubectl delete --grace-period=0 <pod_name>`. Could you try that if the pod gets stuck again?

If `docker inspect` return a 500 error, the pod worker in kubelet will retry the request again. If docker keeps returning the same error, the worker may be stuck and never gets to actual syncing. To work around this, we can perhaps add a retry counter and ignore the container once we've hit a threshold (e.g., fail to inspect the same container for N times). There was a [similar workaround](https://github.com/yujuhong/kubernetes/blob/b1a6ee26efeda0c28825df12e2e4da0a550e08fd/pkg/kubelet/dockertools/manager.go#L2115) that simply fakes an unknown state when docker returns `no such container` even though the container exists in `docker ps`.

> Container doesn't show anywhere

This is a strange case. If the container does not show up anywhere in `docker ps -a`, kubelet should not even try to inspect it. Could you share the relevant logs?

**Comment by andrewmichaelsmith** [comment]
@yujuhong Can confirm that --grace-period=0 will remove the "Terminating" pod. This is useful, thanks.

A retry counter sounds like a good solution to me. 

I had 5 stuck terminating pods today but all appeared in `docker ps -a` I will update if/when I see it again (I'm fairly confident I did see it, but happy to provide logs).

**Comment by yujuhong** [comment]
> @yujuhong Can confirm that --grace-period=0 will remove the "Terminating" pod. This is useful, thanks.

Great.

> A retry counter sounds like a good solution to me.

Filed a bug for this #24361

> I had 5 stuck terminating pods today but all appeared in docker ps -a I will update if/when I see it again (I'm fairly confident I did see it, but happy to provide logs).

That'd be great. I am closing this issue since the root cause is handled through the docker issue, and the workaround will be addressed in the #24361. If you have new information about kubelet inspecting a non-existent container, feel free to reopen the issue!

**Comment by andrewmichaelsmith** [comment]
@yujuhong Brilliant, thanks very much.
